### PR TITLE
A new KeyPressDispatcher supporting function keys

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Triggered by #3827 - a way to have a key press dispatcher that supports both normal keys having a Char value and function keys that have not. Does **not** contain any demo, but I did test as follows (plus exempting F1 from cleanupKeyDispatcher):
```
keyPressDispatcher[Input.Keys.F1] = {game.setScreen(CivilopediaScreen(gameInfo.ruleSet)) }
```
That two-class system dividing world screen keys into permanent and transient ones would have cleaner solutions, too, but I'll keep this small.

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
Do not hesitate expressing like/dislike/horror. e.g. I haven't spent a single brain cell on scoping for the new types.
A note on the WorldScreen change: overloading equality with asymmetric types is a no-go in Kotlin - so no `override operator != (KeyCharAndCode,Char)` or similar, therefore KISS and I did what you see - not even `it != KeyCharAndCode(' ')` as the whole routine needs work anyway

File: core/src/com/unciv/ui/utils/CameraStageBaseScreen.kt [hunk 1]
```diff
@@ -24,36 +24,77 @@ import java.util.HashMap
...
+    constructor(code: Int): this(Char.MIN_VALUE, code)
+    // helper for use in InputListener keyTyped()
+    constructor(event: InputEvent?, character: Char)
+            : this (
+                character.toLowerCase(),
+                if (character == Char.MIN_VALUE && event!=null) event.keyCode else 0
+            )
+}
+
+class KeyPressDispatcher {
```
Comment (yairm210) [inline core/src/com/unciv/ui/utils/CameraStageBaseScreen.kt:41]: We can make `KeyPressDispatcher : HashMap<KeyCharAndCode, (() -> Unit)>`
And then we can save the get/set of the KeyCharAndCode

Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/utils/CameraStageBaseScreen.kt:41]: I know but I preferred this variant when I considered inheritance - maybe I thought something like 'you can explicitly see everything that's going on'. You're probably right. Let's see whether git breaks this as badly as it broke the translation file...

**Comment by SomeTroglodyte** [comment]
Broken by rebase - again

**Comment by yairm210** [comment]
:(

**Comment by SomeTroglodyte** [comment]
Or maybe just the visuals are broken - from here it says 2 files changed - correct. When I click on the tab it says 23 - all the stuff I didn't touch... How can one trust that?. Add to this a local display bug (reproducibly - after rebases and switching branches while leaving a file opened that's touched by those branches it _will_ display outdated file contents until you either restart or just close and reopen the file) - and I can't stop doubting it.

Recreating a commit from local stash *is* easier than amending a branch under the conditions a) it is local+remote and b) master moved forward touching any of the contained files.

I remember (you too probably) I struggled similarly a year ago - but I'm sure I relearned all that quickly. Back then I started having trouble just keeping things in sync locally while still moving master forward keeping up, and that somehow seems easy and transparent now. But once a change is pushed and master moved past it with conflicts - stumped. It should be easy, git exists for that purpose... Hmmm... maybe it's the IDE and I should try the cli instead.

..or go and remerge this instead of chatting...

**Pull request closed without merging**